tmt/options.py

```diff
@@ -317,6 +317,12 @@ def _check_method(self, context: 'tmt.cli.Context', args: List[str]) -> None:
                 # Stop search at the first argument looking like a subcommand
                 elif is_likely_subcommand(arg, subcommands):
                     break
+                # Skip if the first argument is an option of the previous subcommand
```

Comment:
I don't understand what exactly is happening at this point. Could you add a bit more context to illustrate when this situation happens?

Reply:
Hi @happz, without the patch, `tmt run --rm pl disc provin -h local` raise error like:

```bash
Unsupported discover method 'local'.
```

It is because `['provin', '-h', 'local']` is taken as **args** of subcommand of `discover`. Note that `-h` results in `how = local`.
* https://github.com/teemtee/tmt/blob/main/tmt/options.py#L305
  ```python
   301	            for index, arg in enumerate(args):
   302	                # Handle '--how method' or '-h method'
   303	                if arg in ['--how', '-h']:
   304	                    try:
   305	                        how = args[index + 1]
   306	                    except IndexError:
   307	                        pass
   308	                    break
  ```

Hence we ignore `-XXX` in function `_check_method` and let other checking methods validate the options of subcommand. 

With the patch, `tmt run --rm pl disc provin -h local` raise error like:

```
Unsupported discover option 'provin'.
```

Since `provin` is not a valid short name of subcommand, it should be taken as the option of subcommand `disc` (i.e. `discover`).

Reply:
Hmm, so this checks strings between a step name and the actual `-h/--how`? So I can have a plugin that accepts command line arguments (e.g. `report -h email mprchlik@redhat.com tft@redhat.com`), and this check will not treat them as invalid step names, because they do appear *after* `-h`?

Could you add a gist of this as a comment to the code, please? It's a tricky corner case, our future selves will thank us for extra comments :)future

Reply:
No, sorry, I still don't get it :(

What puzzles me is the "the first argument is an option of the previous subcommand". What happens in a case like this: `disc --log-topic key-normalization -h fmf ...`? Walking the list of arguments, the first would be `--log-topic`, we would skip that because it starts with `-`, but then the next token would be `key-normalization`: it's not in `['--how', '-h']`, it doesn't start with `--how=` or `-h`, it's not a subcommand, and it doesn't start with `-`, therefore the only left branch to take is the new one with `GeneralError`.

Reply:
Hi @happz, I have updated the related code according to your comments. Now the patch can just handle with those args which don't starts with `-`. Please help to review again, thanks!

Reply:
Marking as resolved, we seem to be happy with the complicated yet apparently working algorithm.

Reply:
Thank you very much for providing the working algorithm, @happz !